lib/api/controllers/authController.js

```diff
@@ -340,6 +369,34 @@ class AuthController {
 
     return Bluebird.resolve({});
   }
+
+  /**
+   * @param {Request} request
+   */
+  refreshToken(request) {
+    assertIsAuthenticated(this.anonymousId, request);
+
+    const options = {};
+    if (request.input.args.expiresIn) {
+      options.expiresIn = request.input.args.expiresIn;
+    }
+
+    return this.kuzzle.repositories.token.generateToken(request.context.user, request, options)
+      .then(token => {
+        // we allow a grace period before expiring the current token
+        setTimeout(
```

Comment:
After a long and hard think about this "grace period" feature: I thought it was a good idea, but now I'm not so sure.
As you implemented it, the same token can be refreshed multiple times, generating many tokens from a single one. And we can also keep a token alive by constantly refreshing it.
I think that this introduces a security risk.

Still, I find this option interesting.
I think that this can work _only_ if this token cannot be refreshed anymore during the grace period, maybe by adding a flag in the information held by Redis?
This would allow the use of the token during the grace period, while preventing the token leak I described above.

Also, instead of using `expire` with a TTL of `0` (at least I assume that's what the redis client does), you can directly provide the grace period in milliseconds to it (`expire` does not expire a key, but it changes its TTL)

Reply:
Since time is short, I'm continuing this work. As seen together, I'll solve the token bomb problem (even if we both don't know if this is a real risk or not), and also the realtime subscriptions expiring with the 1st token instead of being refreshed too.